File: aoc2025/go/challenges/day7.go
package challenges

import (
	"log/slog"

	"aoc/grid"

	logger "github.com/Bparsons0904/goLogger"
)

func Day7() {
	log := logger.New("Day7")

	tachyonGrid := grid.New("day7.part1")

	timer := log.Timer("Part 1 Timer")
	part1Count := processTachyonBeamSplitCounter(tachyonGrid)
	timer()

	timer = log.Timer("Part 2 Timer")
	part2Count := processTachyonBeamRoutesCounter(tachyonGrid)
	timer()

	log.Info("part1", "Part 1", part1Count, "Part 2", part2Count)
}

type TachyonGraph map[grid.Point]int

func processTachyonBeamRoutesCounter(tachyonGrid *grid.Grid) int {
	techyonGraph := make(TachyonGraph)

	rowLength := len(tachyonGrid.Map[0])
	for i := len(tachyonGrid.Map) - 1; i >= 0; i-- {
		for j := range rowLength {
			if tachyonGrid.Map[i][j] == grid.TACHYON {
				tachyonPathCount := 0
				if j-1 >= 0 {
					tachyonPathCount = techyonGraph.locateTachyon(
						tachyonGrid,
						grid.Point{X: j - 1, Y: i},
					)
				}

				if j+1 < rowLength {
					tachyonPathCount += techyonGraph.locateTachyon(
						tachyonGrid,
						grid.Point{X: j + 1, Y: i},
					)
				}
				techyonGraph[grid.Point{X: j, Y: i}] = tachyonPathCount
			}
		}
	}

	total := techyonGraph.locateTachyon(
		tachyonGrid,
		grid.Point{X: tachyonGrid.Current.X, Y: tachyonGrid.Current.Y},
	)

	return total
}

func (tachyonGraph TachyonGraph) locateTachyon(tachyonGrid *grid.Grid, point grid.Point) int {
	for i := point.Y; i < len(tachyonGrid.Map); i++ {
		value, found := tachyonGraph[grid.Point{X: point.X, Y: i}]
		if found {
			return value
		}
	}

	return 1
}

func processPath(point grid.Point, tachyonGrid *grid.Grid) int {
	if point.Y == len(tachyonGrid.Map)-1 {
		slog.Info("reached end", "point", point)
		return 1
	}

	pathCount := 0
	if tachyonGrid.Map[point.Y+1][point.X] == grid.TACHYON {
		if point.X-1 >= 0 {
			slog.Info("left", "point", point)
			pathCount += processPath(grid.Point{X: point.X - 1, Y: point.Y + 1}, tachyonGrid)
		}

		if point.X+1 < len(tachyonGrid.Map[0]) {
			slog.Info("right", "point", point)
			pathCount += processPath(grid.Point{X: point.X + 1, Y: point.Y + 1}, tachyonGrid)
		}
	} else {
		slog.Info("down", "point", point)
		pathCount += processPath(grid.Point{X: point.X, Y: point.Y + 1}, tachyonGrid)
	}

	return pathCount
}

func processTachyonBeamSplitCounter(tachyonGrid *grid.Grid) int {
	tachyonSplitCounter := 0
	tachyonCurrentLines := make(map[int]bool)
	tachyonCurrentLines[tachyonGrid.Current.X] = true
	for _, row := range tachyonGrid.Map {
		newTachyonCurrentLines := make(map[int]bool)
		for x, space := range row {
			if space == grid.TACHYON && tachyonCurrentLines[x] {
				tachyonSplitCounter++
				newTachyonCurrentLines[x-1] = true
				newTachyonCurrentLines[x+1] = true
				delete(tachyonCurrentLines, x)
			}
		}

		for x := range newTachyonCurrentLines {
			if x < 0 || x >= len(tachyonGrid.Map[0]) {
				continue
			}
			tachyonCurrentLines[x] = true
		}
	}

	return tachyonSplitCounter
}

func trimGrid(tachyonGrid *grid.Grid) {
	for i := 1; i < len(tachyonGrid.Map); i++ {
		hasTachyon := false
		for j := 0; j < len(tachyonGrid.Map[i]); j++ {
			if tachyonGrid.Map[i][j] == grid.TACHYON {
				hasTachyon = true
				break
			}
		}
		if !hasTachyon {
			tachyonGrid.Map = append(tachyonGrid.Map[:i], tachyonGrid.Map[i+1:]...)
			i--
		}
	}
}

File: aoc2025/rust/src/bin/day7.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::collections::HashMap;
use std::time::Instant;

const TACHYON: char = '^';
const START: char = 'S';


#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
struct Point {
    x: i32,
    y: i32,
}

struct Grid {
    width: i32,
    height: i32,
    start: Point,
    map: Vec<Vec<char>>,
}

impl Grid {
    fn new(filename: &str) -> io::Result<Self> {
        let file = File::open(filename)?;
        let reader = io::BufReader::new(file);
        let mut start = Point { x: 0, y: 0 };
        let map: Vec<Vec<char>> = reader
            .lines()
            .enumerate()
            .map(|(y, line)| {
                let l = line.unwrap();
                if let Some(x) = l.find(START) {
                    start = Point { x: x as i32, y: y as i32 };
                }
                l.chars().collect()
            })
            .collect();
        
        let height = map.len() as i32;
        let width = if height > 0 { map[0].len() as i32 } else { 0 };

        Ok(Grid { width, height, start, map })
    }
}

fn main() {
    let grid = Grid::new("files/day7.part1").unwrap();
    
    let part1_count = process_tachyon_beam_split_counter(&grid);
    println!("Day 7, Part 1: {}", part1_count);

    let now = Instant::now();
    let part2_count = process_tachyon_beam_routes_counter(&grid);
    let elapsed = now.elapsed();
    println!("Day 7, Part 2: {}", part2_count);
    println!("Part 2 took: {:.2?}", elapsed);
}

fn process_tachyon_beam_split_counter(grid: &Grid) -> i32 {
    let mut tachyon_split_counter = 0;
    let mut tachyon_current_lines: HashMap<i32, bool> = HashMap::new();
    tachyon_current_lines.insert(grid.start.x, true);

    for row in &grid.map {
        let mut new_tachyon_current_lines = HashMap::new();
        for (x, &space) in row.iter().enumerate() {
            let x = x as i32;
            if space == TACHYON && tachyon_current_lines.contains_key(&x) {
                tachyon_split_counter += 1;
                new_tachyon_current_lines.insert(x - 1, true);
                new_tachyon_current_lines.insert(x + 1, true);
                tachyon_current_lines.remove(&x);
            }
        }

        for (x, _) in new_tachyon_current_lines {
            if x >= 0 && x < grid.width {
                tachyon_current_lines.insert(x, true);
            }
        }
    }
    tachyon_split_counter
}

type TachyonGraph = HashMap<Point, usize>;

fn process_tachyon_beam_routes_counter(grid: &Grid) -> usize {
    let mut tachyon_graph = TachyonGraph::new();
    let row_length = grid.width as usize;

    for i in (0..grid.height as usize).rev() {
        for j in 0..row_length {
            if grid.map[i][j] == TACHYON {
                let mut tachyon_path_count = 0;
                if j > 0 {
                    tachyon_path_count += locate_tachyon(&tachyon_graph, grid, Point { x: j as i32 - 1, y: i as i32 });
                }
                if j + 1 < row_length {
                    tachyon_path_count += locate_tachyon(&tachyon_graph, grid, Point { x: j as i32 + 1, y: i as i32 });
                }
                tachyon_graph.insert(Point { x: j as i32, y: i as i32 }, tachyon_path_count);
            }
        }
    }

    locate_tachyon(&tachyon_graph, grid, grid.start)
}

fn locate_tachyon(tachyon_graph: &TachyonGraph, grid: &Grid, point: Point) -> usize {
    for i in point.y as usize + 1..grid.height as usize {
        if let Some(&value) = tachyon_graph.get(&Point { x: point.x, y: i as i32 }) {
            return value;
        }
    }
    1
}